packages/medusa/src/repositories/product.ts

```diff
@@ -1,21 +1,440 @@
+    })
+
+    if (optionsWithoutRelations.withDeleted) {
+      qb.withDeleted()
+    }
+
+    let entities: Product[]
+    let count = 0
+    if (shouldCount) {
+      const result = await qb.getManyAndCount()
+      entities = result[0]
+      count = result[1]
+    } else {
+      entities = await qb.getMany()
+    }
+
+    return [entities, count]
+  },
+
+  async queryProductsWithIds(
+    entityIds: string[],
+    groupedRelations: { [toplevel: string]: string[] },
+    withDeleted = false,
+    select: (keyof Product)[] = [],
+    order: { [column: string]: "ASC" | "DESC" } = {},
+    where: FindOptionsWhere<Product> = {}
+  ): Promise<Product[]> {
+    const entitiesIdsWithRelations = await Promise.all(
+      Object.entries(groupedRelations).map(async ([toplevel, rels]) => {
+        let querybuilder = this.createQueryBuilder("products")
+
+        if (select && select.length) {
+          querybuilder.select(select.map((f) => `products.${f}`))
+        }
+
+        if (toplevel === "variants") {
+          querybuilder = querybuilder.leftJoinAndSelect(
+            `products.${toplevel}`,
+            toplevel,
+            "variants.deleted_at IS NULL"
+          )
+
+          if (!Object.keys(order).some((key) => key.startsWith("variants"))) {
+            // variant_rank being select false, apply the filter here directly
+            querybuilder.addOrderBy(`${toplevel}.variant_rank`, "ASC")
+          }
+        } else if (toplevel === "categories") {
+          const joinScope =
+            where.categories as FindOptionsWhere<ProductCategory>
+
+          querybuilder = this._applyCategoriesQuery(querybuilder, {
+            alias: "products",
+            categoryAlias: "categories",
+            where: joinScope,
+            joinName: "leftJoinAndSelect",
+          })
+        } else {
+          querybuilder = querybuilder.leftJoinAndSelect(
+            `products.${toplevel}`,
+            toplevel
+          )
+        }
+
+        for (const rel of rels) {
+          const [_, rest] = rel.split(".")
+          if (!rest) {
+            continue
+          }
+
+          querybuilder = querybuilder.leftJoinAndSelect(
+            // Regex matches all '.' except the rightmost
+            rel.replace(/\.(?=[^.]*\.)/g, "__"),
+            // Replace all '.' with '__' to avoid typeorm's automatic aliasing
+            rel.replace(/\./g, "__")
+          )
+        }
+
+        if (withDeleted) {
+          querybuilder = querybuilder
+            .andWhere("products.id IN (:...entitiesIds)", {
+              entitiesIds: entityIds,
+            })
+            .withDeleted()
+        } else {
+          querybuilder = querybuilder.andWhere(
+            "products.deleted_at IS NULL AND products.id IN (:...entitiesIds)",
+            {
+              entitiesIds: entityIds,
+            }
+          )
+        }
+
+        querybuilder.setFindOptions({ relationLoadStrategy: "query" })
```

Comment:
question: Wouldn't this cause issues still?

Reply:
It shouldn't I ve tested with 80 req/s without any issues
<img width="1806" alt="Screenshot 2023-05-15 at 10 48 26" src="https://github.com/medusajs/medusa/assets/25098370/3965e58f-cd81-4bc0-b02c-b0d0e1740704">